sorting/tree_sort.cpp

```diff
@@ -0,0 +1,174 @@
+    node *rotateLeft(node *root) {
+        node *temp = root->right;
+        root->right = temp->left;
+        temp->left = root;
+        temp->RED = root->RED;
+        root->RED = true;
+
+        return temp;
+    }
+
+    // Rotate a left leaning subtree to right orientation
+    node *rotateRight(node *root) {
+        node *temp = root->left;
+        root->left = temp->right;
+        temp->right = root;
+        temp->RED = root->RED;
+        root->RED = true;
+
+        return temp;
+    }
+
+    // Flip colors of children in case both have red links, change parent's link to red
+    void flipColors(node *root) {
+        root->RED = true;
+        root->left->RED = false;
+        root->right->RED = false;
+    }
+
+    node *insert(int val, node *root) {
+        if(root == NULL) return getNode(val, true);
+
+        // Check if the provided value should be inserted in the left or right subtree
+        if(val <= root->data) root->left = insert(val, root->left);
+        if(val > root->data) root->right = insert(val, root->right);
+
+        /* Check if any of the red link subtrees are leaning in the wrong direction
+        *  This is a left-leaning Red-Black BST, so all red-links should be left leaning, by definition.
+        *  For more information about it; refer to wikipedia link at top
+        */ 
+        if(isRed(root->right) && !isRed(root->left)) root = rotateLeft(root);
+        if(isRed(root->left) && !isRed(root->left->left)) root = rotateRight(root);
+        if(isRed(root->left) && isRed(root->right)) flipColors(root);
+
+        return root;
+    }
+
+    // Inorder traversal funcition of BST
+    void inorder(node *root, int arr[], int &index) {
+        if(root == NULL) return;
+
+        inorder(root->left, arr, index);
+        arr[index++] = root->data; // Replacing unsorted elements, with the sorted elements
+        inorder(root->right, arr, index);
+    }
+
+
+    void tree_sort(int arr[], int size) {
+        node *root = NULL; // Root pointer for our BST
+
+        // Inserting all the elements of the array into BST
+        for(int i=0; i<size; i++){
+            root = insert(arr[i], root); // Insertion in a BST is O(log(N))
+        }
+
+        int index = 0; // Index value will be passed to help inorder function to replace elements in a sorted order
+        inorder(root, arr, index); // Time complexity: O(N)
+    
+        // As we are inserting N elements into BST, our time complexity becomes O(N*log(N))
+        // We then also perform a inorder traversal, this brings our time complexity to: O(N+N*log(N)) -> O(N*log(N))
+    }
+}
+
+// Utility function to check if array is sorted
+bool isSorted(int arr[], int size) {
+    for(int i=0; i<size-1; i++) {
+        if(arr[i] > arr[i+1]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Utility function to shuffle array randomly
+void shuffle(int arr[], int n) {
+    srand(time(NULL));
```

Comment:
`nullptr` should be used instead. `NULL` is used in C and not C++.
```suggestion
    srand(time(nullptr));
```